CenterForOpenScience/osf.io
# [PLAT-709] Datacite and Crossref DOI clients

## Purpose

Currently we can only get DOIs through EZID this fix creates a client system so we can retrieve dois and other identifiers through crossref, EZID or datacite clients.

## Changes

- Add Identifier folder with clients for crossref, EZID or datacite.
- Reorganizes views to accommodate the client structure.
- Adds new tests and reorganizes old ones to accommodate new structure.
- Adds Migration to attach url to licenses and give each preprint provider a DOI prefix 

## QA Notes

There is a huge a amount both QA and dev testers should know before they try to test this. Bear with me here!

Here is what you need to do to test Datacite:

1. Make sure the the DATACITE_URL in your local.py file is set to `http://mds.test.datacite.org/`
2. Make sure the DATACITE_PASSWORD and DATACITE_USERNAME are set to a correct password and username, contact Sara or I for one.
3. Create a public project or registration and click `Create DOI` to create a test DOI for that Node.
4. Login to `https://mds.test.datacite.org/datasets?find=ByDoiEquals&form` and check that the metadata for the created DOI is correct, and matches the google doc included with the ticket
5. Make sure old Node's with arks still display them properly.

Here is what you need to do to test Crossref:

1. Make sure the the CROSSREF_DEPOSIT_URL in your local.py file is set to ` 'https://test.crossref.org/servlet/deposit'`
2. Make sure the CROSSREF_PASSWORD and CROSSREF_USERNAME are set to a correct password and username, contact Sara, Erin or I for one.
3. Create a Mailgun account
4. Set CROSSREF_DEPOSITOR_EMAIL to your sandbox Mailgun account name for example `CROSSREF_DEPOSITOR_EMAIL = 'postmaster@sandbox64b04143254c42219fae9180095dcf1b.mailgun.org'`
5. Install ngrok and type `ngrok http 8000` in the terminal to connect your container to the web. It should look like:
<img width="642" alt="screen shot 2018-05-14 at 8 04 50 pm" src="https://user-images.githubusercontent.com/9688518/40029345-1f56e344-57b2-11e8-874a-8252ca4c89d1.png">
Then go back to your mailgun account and forward the emails to the crossref endpoint of the API. It should look like this:
<img width="1440" alt="screen shot 2018-05-14 at 8 02 48 pm" src="https://user-images.githubusercontent.com/9688518/40029379-5c093922-57b2-11e8-94a5-05c6de51f9c9.png">

6. Now when you publish or update a preprint the DOI metadata should be created and it should update if edits to the print are made. You can view preprint metadata at `https://test.crossref.org/servlet/useragent?func=showHome`

## Documentation

I've tried to document what I know in this PR, but we should probably include some more docs later

## Side Effects

None that I know of.

## Ticket

https://openscience.atlassian.net/projects/PLAT/issues/PLAT-802
